### DISCUSSION TITLE
DeleteByQueryOperation executed only on one node

### DISCUSSION THREAD

**Discussion by mic2zar** [ravendb/ravendb#15842] [Q&A] [answered]
I am trying to remove all documents created by index to OutputReduceToCollection after deleting index first.

When I am doing it programmatically like this:
```
await _store.Value.Operations
        .SendAsync(new DeleteByQueryOperation(new IndexQuery
        {
             Query = $"from 'CollectionName'"
         }));
```
or manually deleting all documents in Studio, the collection is removed only on one node and my decision is not replicated to other nodes. Why is that? Can I fix above code to send delete request to all nodes?

**Comment by arekpalinski** [comment, marked as answer]
Hi,

The documents created by map-reduce indexes with the usage of `OutputReduceToCollection` feature are artificial documents. They are created independently by the index on each node and aren't replicated. Please check the following docs: 

https://ravendb.net/docs/article-page/5.4/csharp/studio/database/indexes/create-map-reduce-index#artificial-documents--vs--regular-documents

So yes, it means you need to delete them separately on each node.

**Reply by mic2zar** [reply]
I do understand now the reason, but not the rationale :). 

I still have then the unresolved problem of how could I send DeleteByQueryOperation to every node. DeleteByQueryOperation derives from IOperation and not from IServerOperation preventing me from executing anything like that:
documentStore.Maintenance.Server.ForNode("C").Send(new DeleteByQueryOperation(...));

Any solution?

**Reply by arekpalinski** [reply]
You can force the client to contact a specific node if you create the `DocumentStore` with `DisableTopologyUpdates` convention and set only one node URL. 

```
var store = new DocumentStore
{
    Urls = new[] { "node-specific-url },
    Conventions = new DocumentConventions
    {
        DisableTopologyUpdates = true
    }
}.Initialize()
```

https://ravendb.net/docs/article-page/5.4/csharp/client-api/configuration/conventions#disabletopologyupdates

**Reply by garayx** [reply]
you can also try to get the command from `DeleteByQueryOperation` and send it to selected node:

```
                var re = store.GetRequestExecutor();
                using var ctx = JsonOperationContext.ShortTermSingleUse();
                var op = new DeleteByQueryOperation(new IndexQuery { Query = $"FROM 'Orders'" }, new QueryOperationOptions
                {
                    // let us slow down the operation a bit to make sure we'll get "in-progress" notification before "completed"
                    MaxOpsPerSecond = 500
                });
                var cmd = op.GetCommand(store, store.Conventions, ctx, re.Cache);
                var (currentIndex, currentNode) = await re.GetRequestedNode("A");
                await re.ExecuteAsync(currentNode, currentIndex, ctx, cmd);
```

**Reply by mic2zar** [reply]
Thanks a million. I tried second solution and it worked, but I guess the first one would also make its job. 

Now all the documents are gone on all three nodes. Still, if I call statistic operation I receive an info that collection is there with 0 documents. 
Here is the request:
`var stats = await documentStore.Maintenance.SendAsync(new GetDetailedCollectionStatisticsOperation());`
And here relevant part of the response:
```
{[~DeletedLessons, Raven.Client.Documents.Operations.CollectionDetails]}
Key [string]:"~DeletedCourses"
Value:{Raven.Client.Documents.Operations.CollectionDetails}
CountOfDocuments [long]:0
Name [string]:"~DeletedCourses"
```
No reference to this collection can be found in Studio anymore, but apparently server still knows about it even if both Index and all documents created by OutputCollection are gone. Could you please let me know, what else should I call to get rid of it completely.

**Reply by arekpalinski** [reply]
Under the covers the created structures where we persisted those documents are still there. But they are empty so don't occupy any space.
